certbot/certbot/_internal/plugins/disco.py

```diff
@@ -339,3 +334,69 @@ def __str__(self):
         if not self._plugins:
             return "No plugins"
         return "\n\n".join(str(p_ep) for p_ep in self._plugins.values())
+
+
+def _provides(target_class: Type[interfaces.Plugin], iface: Type) -> bool:
+    if issubclass(target_class, iface):
+        return True
+
+    if iface == interfaces.Plugin and interfaces.IPluginFactory.providedBy(target_class):
+        warnings.warn("Zope interface certbot.interfaces.IPluginFactory is deprecated, "
+                      "use ABC certbot.interface.Plugin instead.")
+        return True
+
+    return False
+
+
+def _implements(target_class: Type[interfaces.Plugin], iface: Type) -> bool:
+    if issubclass(target_class, iface):
+        return True
+
+    if iface == interfaces.Plugin and interfaces.IPlugin.implementedBy(target_class):
+        warnings.warn("Zope interface certbot.interfaces.IPlugin is deprecated, "
+                      "use ABC certbot.interface.Plugin instead.")
+        return True
+
+    if iface == interfaces.Authenticator and interfaces.IAuthenticator.implementedBy(target_class):
+        warnings.warn("Zope interface certbot.interfaces.IAuthenticator is deprecated, "
+                      "use ABC certbot.interface.Authenticator instead.")
+        return True
+
+    if iface == interfaces.Installer and interfaces.IInstaller.implementedBy(target_class):
+        warnings.warn("Zope interface certbot.interfaces.IInstaller is deprecated, "
+                      "use ABC certbot.interface.Installer instead.")
+        return True
+
+    return False
+
+
+def _verify(target_instance: interfaces.Plugin, target_class: Type[interfaces.Plugin],
+            iface: Type) -> bool:
+    if issubclass(target_class, iface):
+        # No need to trigger some verify logic for ABCs: when the object is instantiated,
+        # an error would be raised if implementation is not done properly.
+        # So the checks have been done effectively when the plugin has been initialized.
+        return True
+
+    zope_iface: Optional[Type[zope.interface.Interface]] = None
+
+    if iface == interfaces.Plugin:
+        zope_iface = interfaces.IPlugin
+    if iface == interfaces.Authenticator:
+        zope_iface = interfaces.IAuthenticator
+    if iface == interfaces.Installer:
+        zope_iface = interfaces.IInstaller
+
+    if not zope_iface:
+        raise ValueError(f"Unexpected type: {iface.__name__}")
+
+    try:
+        zope.interface.verify.verifyObject(iface, target_instance)
+        return True
```

Comment:
Before returning here, I think we should probably both call `warnings.warn` and `logger.warning` to tell the user something like:

> `target_class` only implements Certbot's legacy plugin interfaces which will stop working in a future version of Certbot.

Reply:
The warnings emitted by `_implements` and `_provides` are not sufficient in your opinion?

Reply:
I don't think so just because when testing with an authenticator that inherits from `common.Plugin` but is still relying on the `IAuthenticator` zope interface, if you run a command like `certbot certonly -a <plugin name>` no warning seems to be raised.

I also think it'd be good to use `logger.warning` to show a user facing message about this as well to help people see it before their setup breaks.

I'm not exactly sure where these warnings/log messages need to be to ensure these issues are properly flagged. Feel free to try and look through the code and figure that out if you like, but I know that warning here covers the case I just mentioned. My thought now is we may as well add `logging.warning` calls to go with your existing `warnings.warn` calls too. I'd rather make too much noise than too little.

Reply:
I see your point. I agree with you about calling `logging` alongside with `warning`. I think however that adding one more warning logic. The points of the warnings in the `disco` module is about warning that the old interfaces have been used to detect a plugin, not that they have been used at all.

Your use case will be covered by the warnings at the `interface` module that will warn about any use of the old interfaces, and that I will add in one of the follow-up PR.

Reply:
If I'm understanding you and you'd like to just add `logging` calls aside the existing `warning` calls without adding a `warning` call here, my concern is that no user facing warning will be printed/logged in the case I described.

Reply:
One of my sentence has been cut. But basically I was saying that users will receive messages once the `interface` module will get modifications to emit the deprecation messages when the old interfaces are used.

I think also that at that point, the messages here may be useless since users and developers will be warned for every cases where the old interfaces are involved.

Reply:
And I am also saying that the message you are proposing since the beginning of this conversation is not relevant since it is not representing what it is happening and what we are supposed to warn in the `disco` module. We are trying to say here that plugins must not be detected by the implementation of the old interfaces but by the new interfaces. In that specific part you are proposing the plugins have already been detected, and we are validating the correctness of the interface implementation. Warnings have already been emitted during the detection of the plugins. Also in the use case you described, the plugin has been detected by its implementation of the new interface, since it extends `common.Plugin`, and happens to also implement the old interface, but this is a noop for the plugin discovery purpose.

Also emitting warnings when the old interfaces are used in any context (including the use case you describe) is relevant for sure, but it is a matter of modifying the `interfaces` module to trigger that at the import, not of the `disco` module during plugin détection.

Reply:
Your comment reads a little unnecessarily combative to me, but here are the places I think we may be seeing and thinking about things differently:

1. I am differentiating between error messages for users and those for developers. I think warnings are for developers and log messages are for users. I also think logging messages for users about imported code seems odd to me but maybe that's what you're proposing we do here?
2. I also don't think it's true that warnings have already been emitted during the detection of plugins in the case I describe. You can hit this case with our own plugins such as when using an older version of our Apache or nginx plugin. If you apply [this patch](https://gist.github.com/bmw/1ff59fe13a9974c59d951b8225d4223b) and run `certbot certonly -a nginx ...`, no message is printed. If we later apply a [patch like this](https://gist.github.com/a6529e04afcd9f7aaa98748c1782be6b) (which maybe isn't what you want to do?), that same command blows up.

Does this comment help at all? I just think that if a command that previously worked is going to start failing, we should make sure user facing warnings are printed beforehand. How do you see us avoiding this problem?

Reply:
For 1., I basically agree with you that we should emit both a warning for devs and a logging message for users and it is what I added in my last commit.

For 2., again not with the purpose of these warnings to be emitted when plugins are detected by the old interfaces: since the plugin class extends `common.Installer` it gets the new interfaces from inheritance and is detected by them, and so no warnings are emitted.

OK I think we are running in circle here. I propose a different approach. Let's forget about the warning stuff in the `dico` module, and implement right now the warnings in the `interface` module, which was what I was planning to do after this PR. This way warnings will be emitted in all appropriate situations, including the bare usage of the old interfaces even when they are not involved for plugin detection. We were planning to remove them anyway.

What do you think of this ?

Reply:
After a discussion with @bmw, we agreed that in some usecases, plugins may be verified without initialization, and plugins may be initialized but not verified, due to the complexity of the plugins management system. So it is better warn on all situations, at the cost of duplicated warnings in the other situations.